### PR TITLE
Eric/new light control and simultaneous light training

### PR DESCRIPTION
**Pull request by nazric**
Lots of changes happening here. There are a few changes to the nets thought shouldn't have any consequence on other people's work, but I can go in and remove them as they aren't strictly necessary, and should probably be in a different PR.

This allows us to train faster, and with more flexibility. It should be good to go for training on either PoundSign or VitalTri. 

If @guevarrm has any changes to the reward function or observation function copy it and add it to TrainingGym/env/reward_functions.md once you branch off these latest changes.

Now there is also support to log messages to file (TrainingGym/training_log.txt), as the console gets drowned in stable baselines and sumo output. 

Also take a look in global_config and training_config for new options for training.

### CODE REVIEW

**Comment by Sahmwell** [comment]
This one also has conflicts now that https://github.com/Sahmwell/G15_Capstone/pull/20 has been merged in. I'd recommend rebasing this branch on top of https://github.com/Sahmwell/G15_Capstone/pull/18/ as well so that I can merge these both in at the same time since I think these two branches have conflicting changes.

**Comment by nazric** [comment]
@Sahmwell should be good for merge now. I added the changes from the other PR referenced above.

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [inline TrainingGym/env/SumoEnvParallel.py:36]: I think this could just be `for i in range(len(controlled_lights))` but functionally I suppose it doesn't make a difference

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [inline TrainingGym/env/SumoEnvParallel.py:90]: love me some defaultdicts 👌

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [inline TrainingGym/env/SumoEnvParallel.py]: not something that needs to change in this PR, but this can be simplified to `if model['model']:`

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [inline TrainingGym/env/SumoEnvParallel.py:171]: might be worth it iterate over the important roads rather than the vehicles, since at any instance the majority of vehicles probably won't be on an important road, I think it should allow for less calls to traci each time step.

something like this I believe should do the same thing with fewer traci calls:
```python
    def _get_sumo_values(self):
        vehicles_on_edge = defaultdict(lambda: [])
        for edge in all_important_roads:
            vehicles_on_edge = self.sumo.edge.getLastStepVehicleIDs(edge)
            for v in vehicles_on_edge:
                v_pos = np.array(self.sumo.vehicle.getPosition(v))
                vehicles_on_edge[edge].append(
                    {"name": v, "wait_time": self.sumo.vehicle.getWaitingTime(v), "pos": v_pos}
                )
        return vehicles_on_edge
```

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [inline TrainingGym/env/SumoEnvParallel.py:240]: We can avoid this disclaimer if we call `self._get_sumo_values()` in this method, since this is the only place that uses `vehicles_on_edge` which I think is the only thing that `self._get_sumo_values()` populates.

This should also allow us to remove the `self.vehicles_on_edge` property, and remove it from the initialization/reset functions.

File: TrainingGym/test.py
Comment (Sahmwell) [inline TrainingGym/test.py:21]: Same thing as above where I think this could just be `for i in range(len(controlled_lights))`

File: TrainingGym/test.py
Comment (Sahmwell) [inline TrainingGym/test.py:34]:
```python
while not done:
```
 is probably a little nicer

File: TrainingGym/test.py
Comment (Sahmwell) [inline TrainingGym/test.py:42]: I'm pretty sure this `while` loop doesn't terminate now that the `break` is removed

File: TrainingGym/train.py
Comment (Sahmwell) [inline TrainingGym/train.py:28]: please `for i in range(len(controlled_lights))`

File: TrainingGym/train.py
Comment (Sahmwell) [inline TrainingGym/train.py]: I think this can be `for learning_light in controlled_lights:`

File: TrainingGym/train.py
Comment (Sahmwell) [inline TrainingGym/train.py]: I don't think this should be required. Consider using a different name for the variable instead of `env` since I think `env` is being persisted due to our library `env`. If we do that it should be automatically deleted once we exit this function

File: TrainingGym/train.py
Comment (Sahmwell) [inline TrainingGym/train.py]: we can just return right here

`return  DummyVecEnv([lambda: SumoEnvParallel(steps_per_episode, visualize_training, node_name)])`

This has the added bonus of not overwriting our `env` library variable

File: TrainingGym/train.py
Comment (Sahmwell) [inline TrainingGym/train.py]: same here

**Review by Sahmwell — COMMENTED** [review]
I provided some feedback, it's looking mostly good but there are some changes that I think should be addressed before we merge (biggest one for me was the non terminating while loop in the test.py file)

**Review by Sahmwell — CHANGES_REQUESTED** [review]
Forgot to finish the review with the comment marked as request changes, so this'll be the catch-all request changes comment

File: TrainingGym/env/SumoEnvParallel.py
Comment (nazric) [reply TrainingGym/env/SumoEnvParallel.py:36]: I'm not sure, since we're deleting items from the list the index will change, so we have to traverse it backwards. Or at least that's what I thought. maybe python is different/

File: TrainingGym/env/SumoEnvParallel.py
Comment (nazric) [reply TrainingGym/env/SumoEnvParallel.py:90]: Yeah they're pretty sweet ngl

File: TrainingGym/env/SumoEnvParallel.py
Comment (Sahmwell) [reply TrainingGym/env/SumoEnvParallel.py:36]: Ah yeah, you're right.

**Review by Sahmwell — APPROVED** [review]

**Comment by Sahmwell** [comment]
Merged, I think it's still worth trying to resolve the other comments, but I'm not going to block this PR over it

**Pull request merged**
